rs/types/types/src/batch/ingress.rs

```diff
@@ -1,177 +1,141 @@
-                .id_and_pos
-                .iter()
-                .map(|ingress_offset| {
-                    Ok((
-                        IngressMessageId::new(
-                            Time::from_nanos_since_unix_epoch(ingress_offset.expiry),
-                            MessageId::try_from(ingress_offset.message_id.as_slice())?,
-                        ),
-                        ingress_offset.offset,
-                    ))
-                })
-                .collect::<Result<Vec<_>, ProxyDecodeError>>()?,
-            buffer: payload.buffer,
+            serialized_ingress_messages,
         })
     }
 }
 
-/// Index of an ingress message in the IngressPayload.
-type IngressIndex = usize;
+#[derive(Debug, PartialEq)]
+pub struct IngressPayloadError(HttpRequestError);
 
-/// Position of serialized ingress message in the payload buffer.
-type BufferPosition = u64;
-
-#[derive(Eq, PartialEq, Debug)]
-/// Possible errors when accessing messages in an [`IngressPayload`].
-pub enum IngressPayloadError {
-    IndexOutOfBound(IngressIndex),
-    IngressPositionOutOfBound(IngressIndex, BufferPosition),
-    DeserializationFailure(String),
-    MismatchedMessageIdAtIndex(IngressIndex),
+impl Display for IngressPayloadError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        self.0.fmt(f)
+    }
 }
 
 impl IngressPayload {
     /// Return the number of ingress messages contained in this payload
     pub fn message_count(&self) -> usize {
-        self.id_and_pos.len()
+        self.serialized_ingress_messages.len()
     }
 
-    /// Return all MessageIds in the payload.
-    pub fn message_ids(&self) -> Vec<IngressMessageId> {
-        self.id_and_pos
-            .iter()
-            .map(|(id, _)| id.clone())
-            .collect::<Vec<_>>()
+    /// Return all [`IngressMessageId`]s in the payload.
+    pub fn message_ids(&self) -> impl Iterator<Item = &IngressMessageId> {
+        self.serialized_ingress_messages.keys()
     }
 
     /// Return true if the payload is empty.
     pub fn is_empty(&self) -> bool {
-        self.id_and_pos.is_empty()
+        self.serialized_ingress_messages.is_empty()
     }
 
-    // TODO(kpop): run some benchmarks and see if it makes sense to change the type of
-    // `[IngressPayload::id_and_pos]`
-    pub fn get_by_id(&self, ingress_message_id: &IngressMessageId) -> Option<SignedIngress> {
-        let (index, _) = self
-            .id_and_pos
-            .iter()
-            .enumerate()
-            .find(|(_, (id, _))| id == ingress_message_id)?;
-
-        self.get(index)
-            .map(|(_, ingress_message)| ingress_message)
-            .ok()
+    /// Return the [`SignedIngress`] referenced by the [`IngressMessageId`].
+    /// Return [`IngressPayloadError`] if we fail to deserialize the message.
+    pub fn get_by_id(
+        &self,
+        ingress_message_id: &IngressMessageId,
+    ) -> Option<Result<SignedIngress, IngressPayloadError>> {
```

Comment:
NIT:

I find `Option<Result<_, _>>` signatures rather strange. Maybe change it to a `Result<Option<_>, _>` using `.transpose()`? 

I mean you only use it in one place so its probably fine, it just looks a but smoother the other way around.